awx/main/dispatch/worker/base.py

```diff
@@ -0,0 +1,163 @@
+# Copyright (c) 2018 Ansible by Red Hat
+# All Rights Reserved.
+
+import os
+import logging
+import signal
+import sys
+from uuid import UUID
+from Queue import Empty as QueueEmpty
+
+from kombu import Producer
+from kombu.mixins import ConsumerMixin
+
+from awx.main.dispatch.pool import WorkerPool
+
+logger = logging.getLogger('awx.main.dispatch')
+
+
+def signame(sig):
+    return dict(
+        (k, v) for v, k in signal.__dict__.items()
+        if v.startswith('SIG') and not v.startswith('SIG_')
+    )[sig]
+
+
+class WorkerSignalHandler:
+
+    def __init__(self):
+        self.kill_now = False
+        signal.signal(signal.SIGINT, self.exit_gracefully)
+
+    def exit_gracefully(self, *args, **kwargs):
+        self.kill_now = True
+
+
+class RoundRobinStrategy(object):
+    '''
+    Distribute messages based on int(UUID) % total_workers
+    This is used for callback event delivery, where we want to dispatch *as
+    fast as possible*, but don't care as much about even distribution amongst
+    the worker processes (because each event is handle very quickly).
+    '''
+
+    def schedule(self, consumer, body, message):
+        if "uuid" in body and body['uuid']:
+            try:
+                queue = UUID(body['uuid']).int % len(consumer.pool)
+            except Exception:
+                queue = consumer.total_messages % len(consumer.pool)
+        else:
+            queue = consumer.total_messages % len(consumer.pool)
+        return queue
+
+
+class LeastBusyStrategy(object):
+    '''
+    Attempt to deliver messages to the *least busy* worker (the worker with the
+    shortest work queue).
+    This is used for task dispatch, where we want to schedule *evenly* to run
+    as many simultaneous tasks as possible (ideally, each worker _never_ has
+    a message backlog).
+    '''
+
+    def schedule(self, consumer, body, message):
+        queue = 0
+        size = sys.maxint
+        for i, worker in enumerate(consumer.pool.workers):
+            if len(worker.managed_tasks) < size:
+                queue = i
+                size = len(worker.managed_tasks)
+        return queue
+
+
+class AWXConsumer(ConsumerMixin):
+
+    def __init__(self, name, connection, worker, queues=[], pool=None, strategy=RoundRobinStrategy):
```

Comment:
...much like the callback receiver implementation in 3.3.0 (on which this code is based), this entry point is a `kombu.ConsumerMixin`.  This interface lets us define a set of queues we care about (the global default queue, the instance's hostname-specific queue, and the broadcast queue), fetch new events from those queues, and write them to the pool of workers using a strategy.

There are two strategies defined in this module:

1.  `RoundRobinStrategy` matches the behavior of the 3.3.0 callback receiver.  The goal is to pass events to the pool of workers as *fast* as possible.  We don't care much about even distribution, because the handling of each task (saving an event to the database) is super fast.

2.  `LeastBusyStrategy` - this is the strategy used for task execution.  The idea is that we should give new tasks to the process that is the *least* busy - in this way, we aim to ensure that each worker is as busy as possible (we don't want Worker #1 to have zero tasks and Worker #2 to have a queue with 10 tasks sitting in it, because a long running task could hold things up).